### ISSUE TITLE
Padding below X axis?

### ISSUE THREAD

**Issue by ak2344** [gonum/plot#294]
I've added custom labels to my X axis but the text is getting cut off (the bottom half of the "y" in "May" isn't visible). Is there a way to add padding to the bottom of the graph?

**Comment by ctessum** [comment]
You can use draw.Crop: https://godoc.org/github.com/gonum/plot/vg/draw#Crop.

However, if this is happening it is probably a bug so if you are willing to submit an example that reproduces the problem that would be appreciated.

**Comment by ak2344** [comment]
Thanks!

I will put together an example. What information would you need?

**Comment by ak2344** [comment]
Attaching an example of a graph.
![vs65uphi6xjmv47u](https://cloud.githubusercontent.com/assets/1429953/16716490/c13641f4-46cb-11e6-8b3e-f21b5a53871a.png)

**Comment by kortschak** [comment]
We'd need the code that generated it too.

**Comment by danp** [comment]
Just ran into this myself. Taking the [bar charts example](https://github.com/gonum/plot/wiki/Example-plots#bar-charts) and changing the last NominalX name to "Twenty" produces:

![barchart](https://user-images.githubusercontent.com/2182/97818744-9ec17b00-1c7a-11eb-951c-82122580378a.png)

This is with `gonum.org/v1/plot v0.8.0` in go.mod.

Will see if I can fix it up with draw.Crop.

**Comment by kortschak** [comment]
Thanks. I'll have a look at this soon.

**Comment by kortschak** [comment]
It looks to me like the glyphboxes are not being drawn on the tick labels in the tests either. They should be.

**Comment by kortschak** [comment]
I think this fixes the issue.
```diff
diff --git a/axis.go b/axis.go
index 1190ae5..1daf3d9 100644
--- a/axis.go
+++ b/axis.go
@@ -270,12 +270,13 @@ func (a horizontalAxis) draw(c draw.Canvas) {
 
        marks := a.Tick.Marker.Ticks(a.Min, a.Max)
        ticklabelheight := tickLabelHeight(a.Tick.Label, marks)
+       descent := a.Tick.Label.Font.Extents().Descent
        for _, t := range marks {
                x := c.X(a.Norm(t.Value))
                if !c.ContainsX(x) || t.IsMinor() {
                        continue
                }
-               c.FillText(a.Tick.Label, vg.Point{X: x, Y: y + ticklabelheight}, t.Label)
+               c.FillText(a.Tick.Label, vg.Point{X: x, Y: y + ticklabelheight - descent}, t.Label)
        }
 
        if len(marks) > 0 {
diff --git a/vg/draw/text.go b/vg/draw/text.go
index 35b6699..e19ff3e 100644
--- a/vg/draw/text.go
+++ b/vg/draw/text.go
@@ -74,7 +74,7 @@ func (sty TextStyle) Height(txt string) vg.Length {
                return vg.Length(0)
        }
        e := sty.Font.Extents()
-       return e.Height*vg.Length(nl-1) + e.Ascent
+       return e.Height*vg.Length(nl-1) + e.Ascent - e.Descent
 }
 
 // Rectangle returns a rectangle giving the bounds of
```

**Comment by kortschak** [comment]
A number of tests in vg/... don't update with the `-regen` flag.

**Comment by kortschak** [comment]
No, it's more complicated than that; the wiring of descents into font consideration is quite _ad hoc_.

**Comment by danp** [comment]
Thank you!

**Comment by kortschak** [comment]
@danp Thanks for following up with a reproducer; that's what go this fixed.

**Comment by wcshds** [comment]
@kortschak Some characters (such as ÀÉÈÎÙÔ) is also cut off in title and y axis label.
![test](https://user-images.githubusercontent.com/8537285/165152235-89f4c42d-bd05-496b-85d2-e8cbff7a8155.png)

**Comment by kortschak** [comment]
Thanks @wcshds. Maybe we should grow the bounding box a little.

@sbinet WDYT?

**Comment by sbinet** [comment]
apologies for dropping this on the floor.
I'll have a look.

**Comment by sbinet** [comment]
ok, so the bounding box is indeed completely missing the accents:
![glyphbox_accents](https://user-images.githubusercontent.com/1070920/168066938-90c31def-fef8-449f-b496-89cda9b67d16.png)

asking on the `#ui-text-rendering` channel, Egon reminded me of the `x/image/font.BoundString` function that does what we want (and removing code from gonum to replace it with code from the almost-stdlib is great).
but there are discrepancies.
the code in `gonum/plot` - which more or less does what `font.BoundString` does - returns twice the width of `font.BoundString`:

https://go.dev/play/p/_hvN0Lw5YJ6

the `scale` parameter of `opentype.Face` (`font.BoundString` operates on `font.Face`s) is: `Int26_6(0.5 + (fntSize * imgDPI * 64)/72)) = 1024` (with `fntSize=12` and `imgDPI=96`)
the `pixelsPerEm` parameter passed to `sfnt.Font.GlyphAdvance` (in `doGonumWidth`) equals to `2048`. and this value is taken straight from `sfnt.Font.UnitsPerEm()`
`font.BoundString` just uses `sfnt.Font.GlyphBounds` (whose code is like `sfnt.Font.GlyphAdvance` as far as `advance` is concerned).
so... clearly there is a logical flaw somewhere. but I am stuck.

**Comment by kortschak** [comment]
Do you want something instead like this https://go.dev/play/p/GTsNZ7P8brW?

The `font.Face.GlyphBounds` and `sfnt.Font.GlyphAdvance` look like they have different understanding of advance.

**Comment by sbinet** [comment]
ok...
so I gave up on trying to understand where exactly the difference is coming from.
- `sfnt.Font.GlyphBounds` [here](https://github.com/golang/image/blob/master/font/sfnt/sfnt.go#L1551..L1602)
- `sfnt.Font.GlyphAdvance` [here](https://github.com/golang/image/blob/master/font/sfnt/sfnt.go#L1604...L1635)
ie: as far as `advance` is concerned, they look the same to me.

anyways. dropping the use of `plot/font.Face.Extents()` for the height, and using instead `x/image/font.BoundString`, I get the following image (using this new scheme *only* for the `Ôag` and `Oag` labels (bottom-left quarter)):
![accents-1](https://user-images.githubusercontent.com/1070920/169300433-d2ac690c-152f-4614-8929-3a9c3e71401b.png)

using that scheme for all strings:
![accents-2](https://user-images.githubusercontent.com/1070920/169300501-7a1a303b-198b-471f-9dfd-c7519c316af8.png)

_ie:_ probably have to do with some impedance mismatch between code that uses `plot/font.FaceExtents` directly to get the height of the box and code that uses `plot/text.Handler.Box`.

I'll probably have to modify the `plot/text.Handler` interface, namely that `Box` method that needs to take an additional parameter (the `DPI`).

I must say, I am also considering just switching to [go-text/typesetting](https://github.com/go-text/typesetting) and its [shaping.Shaper](https://pkg.go.dev/github.com/go-text/typesetting@v0.0.0-20220411150340-35994bc27a7b/shaping#Shaper) interface.
this quite large dependency is what Gio and Fyne are using for text typesetting. (and it's still being maintained, while `x/image/font` is less so).
